chainer/optimizer_hooks/gradient_hard_clipping.py

```diff
@@ -47,4 +48,16 @@ def __call__(self, rule, param):
             return
         with chainer.using_device(param.device):
             xp = param.device.xp
-            xp.clip(grad, self.lower_bound, self.upper_bound, out=grad)
+            if xp is backend.chainerx:
+                # TODO(kshitij12345): remove when chainerx.clip supports
+                # kwarg `out`.
+                param.grad = xp.clip(grad, self.lower_bound, self.upper_bound)
+            else:
+                if isinstance(grad, backend.intel64.mdarray):
+                    # Since np.clip doesn't support mdarray
+                    param.grad[param.grad <
+                               self.lower_bound] = self.lower_bound
+                    param.grad[param.grad >
+                               self.upper_bound] = self.upper_bound
+                    return
```

Comment:
When `param.grad` is of type `intel64.mdarray`, `xp` is `numpy` and `np.clip` doesn't ofcourse accept `mdarray`.
So the above hack.

This works. But not sure if this is the correct way to go for `mdarray`.

@asi1024 Your thoughts?

Reply:
How about `param.grad[:] = param.grad.clip(self.lower_bound, self.upper_bound)`?

Reply:
If I remember correctly, I did `dir(param.grad)` to check if the `mdarray` has a method for `clip`, however I didn't see it. 

Will confirm and let you know.

Reply:
I am sorry. You were correct. 
Also it turns out 
`param.grad = param.grad.clip(self.lower_bound, self.upper_bound)` 
is uniform for all backends.

Thank You.